File: lib-segment/src/main/java/websiteschema/mpsegment/tools/StringWordConverter.java
package websiteschema.mpsegment.tools;

import websiteschema.mpsegment.concept.Concept;
import websiteschema.mpsegment.concept.ConceptRepository;
import websiteschema.mpsegment.dict.IWord;
import websiteschema.mpsegment.dict.POS;
import websiteschema.mpsegment.dict.POSArray;
import websiteschema.mpsegment.dict.WordImpl;

import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class StringWordConverter {

    private ConceptRepository conceptRepository = null;
    private static Pattern patternWord = Pattern.compile("\"(.*)\"\\s*=\\s*(\\{.*\\})");
    private static Pattern patternPair = Pattern.compile("([\\w\\d]+)\\s*:\\s*((\\{[^\\}]+?\\})|([\\w\\d]+)|(\\[[^\\]]+?\\]))\\s*(,|$)");
    private static String domainTypeKey = "domainType";
    private static String posTableKey = "POSTable";
    private static String conceptsKey = "concepts";

    public void setConceptRepository(ConceptRepository conceptRepository) {
        this.conceptRepository = conceptRepository;
    }

    public IWord convert(String str) {
        Matcher m = patternWord.matcher(str);
        if (m.matches()) {
            String wordName = m.group(1);
            Map<String, String> properties = convertToMap(m.group(2));
            try {
                wordName = new String(java.net.URLDecoder.decode(wordName, "utf-8"));
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
            }

            WordImpl word = new WordImpl(wordName);
            setDomainType(properties.get(domainTypeKey), word);
            setPOSTable(properties.get(posTableKey), word);
            setConcepts(properties.get(conceptsKey), word);
            properties = null;
            return word;
        }
        return null;
    }

    private void setConcepts(String conceptsString, WordImpl word) {
        List<String> conceptStrList = convertToList(conceptsString);
        if (null != conceptStrList) {
            Concept[] concepts = new Concept[conceptStrList.size()];
            for (int i = 0; i < conceptStrList.size(); i++) {
                String conceptName = conceptStrList.get(i);
                Concept concept = conceptRepository.getConceptByName(conceptName);
                concepts[i] = concept;
            }
            word.setConcepts(concepts);
        }
    }

    private void setPOSTable(String posTableString, WordImpl word) {
        if (null != posTableString) {
            Map<String, String> posTable = convertToMap(posTableString);
            POSArray posArray = new POSArray();
            for (String pos : posTable.keySet()) {
                posArray.add(new POS(pos, Integer.parseInt(posTable.get(pos))));
            }
            posArray.buildPOSArray();
            word.setPosArray(posArray);
            posTable = null;
        }
    }

    private void setDomainType(String domainType, WordImpl word) {
        if (null != domainType) {
            word.setDomainType(Integer.parseInt(domainType));
        }
    }

    private Map<String, String> convertToMap(String str) {
        if (str.matches("\\{.*\\}")) {
            String content = str.substring(1, str.length() - 1);
            Matcher m = patternPair.matcher(content);
            Map<String, String> map = new LinkedHashMap<String, String>();
            int start = 0;
            while (m.find(start)) {
                String pair = m.group();
                map.put(m.group(1), m.group(2));
                start += pair.length();
            }
            return map;
        }
        return null;
    }

    private List<String> convertToList(String str) {
        if (null != str && str.matches("\\[.*\\]")) {
            String content = str.substring(1, str.length() - 1);
            String[] array = content.split(",");
            List<String> list = new ArrayList<String>(array.length);
            for (String ele : array) {
                list.add(ele.trim());
            }
            return list;
        }
        return null;
    }

}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/tools/StringWordConverter.scala
package websiteschema.mpsegment.tools

import websiteschema.mpsegment.concept.Concept
import websiteschema.mpsegment.concept.ConceptRepository
import websiteschema.mpsegment.dict.IWord
import websiteschema.mpsegment.dict.POS
import websiteschema.mpsegment.dict.POSArray
import websiteschema.mpsegment.dict.WordImpl
import scala.util.parsing.json.JSON

class StringWordConverter {

  private var conceptRepository: ConceptRepository = null
  private val domainTypeKey = "domainType"
  private val posTableKey = "POSTable"
  private val conceptsKey = "concepts"

  def setConceptRepository(conceptRepository: ConceptRepository) {
    this.conceptRepository = conceptRepository
  }

  def convertJSON(str: String): IWord = {
    JSON.parseFull(str) match {
      case Some(map: Map[String, Any]) => {
        val word = new WordImpl(map("word").asInstanceOf[String])
        word.setDomainType(map(domainTypeKey).asInstanceOf[Double].toInt)
        updatePOSTable(map(posTableKey).asInstanceOf[Map[String, Double]], word)
        val concepts = map.getOrElse(conceptsKey, List[String]()).asInstanceOf[List[String]]
        updateConcepts(concepts, word)
        word
      }
      case _ => null
    }
  }

  def convert(wordStr: String): IWord = {
    val wordNamePattern = """"word"\s*:\s*"(.+?)"\s*,""".r
    wordNamePattern findFirstIn wordStr match {
      case Some(wordNamePattern(wordName)) => {
        val word = new WordImpl(escape(wordName))

        parseDomainType(wordStr, word)
        parsePOS(wordStr, word)
        parseConcepts(wordStr, word)

        word
      }
      case None => throw new RuntimeException(s"Cannot convert to word: $wordStr")
    }
  }

  def escape(word:String) = word.
    replaceAll("""\\r""", "\r").
    replaceAll("""\\n""", "\n").
    replaceAll("""\\/""", "/").
    replaceAll("""\\"""", "\"").
    replace("""\\""", "\\")

  def parseConcepts(wordStr: String, word: WordImpl) {
    val conceptsPattern = """"concepts"\s*:\s*\[(("([^"]+)",?)+)\]\s*,?""".r
    conceptsPattern findFirstMatchIn wordStr match {
      case Some(m) => {
        val concepts = m.group(1)
        val conceptPattern = """"([^"]+)\",?""".r
        val conceptArray = conceptPattern.findAllMatchIn(concepts).
          map(m =>conceptRepository.getConceptByName(m.group(1))).toArray
        word.setConcepts(conceptArray)
        }
      case None => word.setConcepts(Array())
    }
  }

  def parseDomainType(wordStr: String, word: WordImpl) {
    val domainTypePattern = """"domainType"\s*:\s*(\d+)\s*,?""".r
    domainTypePattern findFirstIn wordStr match {
      case Some(domainTypePattern(domainType)) => word.setDomainType(domainType.toInt)
      case None => word.setDomainType(0)
    }
  }

  def parsePOS(wordStr: String, word: WordImpl) {
    val posTablePattern = """"POSTable"\s*:\s*\{(("(\w+)"\s*:\s*(\d+)\s*,?\s*)+)\}\s*,?""".r
    posTablePattern findFirstMatchIn wordStr match {
      case Some(m) => {
        val posTable = m.group(1)
        val posArray = new POSArray()

        val posAndFreq = """"(\w+)"\s*:\s*(\d+)""".r
        posAndFreq.findAllMatchIn(posTable)
          .foreach(m => posArray.add(POS(m.group(1), m.group(2).toInt)))
        posArray.buildPOSArray()
        word.setPosArray(posArray)
      }
      case None =>
    }
  }

  def updateConcepts(conceptList: List[String], word: WordImpl) {
    val concepts = new Array[Concept](conceptList.size)
    for (i <- 0 until conceptList.size) {
      val conceptName = conceptList(i)
      try {
        val concept = conceptRepository.getConceptByName(conceptName)
        concepts(i) = concept
      } catch {
        case _: Exception =>
          println(conceptName)
      }
    }
    word.setConcepts(concepts)
  }

  def updatePOSTable(posTable: collection.immutable.Map[String, Double], word: WordImpl) {
    val posArray = new POSArray()
    for (pos <- posTable.keys) {
      posArray.add(POS(pos, posTable(pos).toInt))
    }
    posArray.buildPOSArray()
    word.setPosArray(posArray)
  }

}
